### PR TITLE
Bazaar Overview Card with random or latest projects

### PR DESCRIPTION
**Pull request by fridajac**
## Hey, I just made a Pull Request!

I have created "BazaarOverviewCard" that can be added to Backstage homepage. It displays a chosen number of either latest or random Bazaar projects. The idea is that it will give users of Backstage a quick way into Bazaar projects in a fun way. 

Changes:
* Backend: Added a "getLatestProjects"-endpoint in the "bazaarApi" that takes a limit of projects as props.
* Frontend: Created "BazaarOverviewCard" that takes the type of card (latest or random) and the limit of projects as props. The card displays the projects and a link to Bazaar.

overviewCard

#### :heavy_check_mark: Checklist

- [X] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [X] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [X] Screenshots attached (for UI changes)
- [X] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/bazaar/src/components/BazaarOverviewCard/BazaarOverviewCard.tsx [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        break;
+      default:
+        response = await bazaarApi.getLatestProjects(limit);
+    }
+
+    let dbProjects: BazaarProject[] = [];
+
+    response.data.forEach((project: any) => {
+      dbProjects.push(parseBazaarProject(project));
+    });
```
Comment (freben) [inline plugins/bazaar/src/components/BazaarOverviewCard/BazaarOverviewCard.tsx:81]: could also do `let dbProjects = response.data.map(parseBazaarProject);`

File: plugins/bazaar-backend/src/service/router.ts [hunk 2]
```diff
@@ -89,6 +89,13 @@ export async function createRouter(
     response.json({ status: 'ok', data: data });
   });
 
+  router.get('/projects/:limit/latest', async (request, response) => {
```
[the remark this answers is not in the document]
Comment (freben) [reply plugins/bazaar-backend/src/service/router.ts:92]: postgres can actually do `order by random()` or something like that i think

File: plugins/bazaar-backend/src/service/router.ts [hunk 3]
```diff
@@ -83,8 +83,14 @@ export async function createRouter(
     response.json({ status: 'ok', data: data });
   });
 
-  router.get('/projects', async (_, response) => {
-    const data = await dbHandler.getProjects();
+  router.get('/projects/:limit?/:order?', async (request, response) => {
```
Comment (freben) [inline plugins/bazaar-backend/src/service/router.ts:86]: Hm is this the right syntax? I thought maybe just `/projects`, and then on the line below do e.g. `const limit = request.query.limit?.toString()` and then the limit will be a string or an undefined. You'd then call the endpoint as `curl .../projects?limit=10`

File: plugins/bazaar/src/api.ts [hunk 4]
```diff
@@ -148,12 +148,15 @@ export class BazaarClient implements BazaarApi {
     );
   }
 
-  async getProjects(): Promise<any> {
+  async getProjects(limit?: number, order?: string): Promise<any> {
     const baseUrl = await this.discoveryApi.getBaseUrl('bazaar');
+    const url =
+      order && limit
+        ? `projects/${encodeURIComponent(limit)}/${encodeURIComponent(order)}`
```
Comment (freben) [inline plugins/bazaar/src/api.ts:155]: yeah as per above you could do

```ts
const query = new URLSearchParams({
  ...(limit ? { limit } : {}),
  ...(order ? { order } : {}),
});
const url = `projects?${query.toString()}`
```

or something along those lines

File: plugins/bazaar/src/api.ts [hunk 5]
```diff
@@ -148,12 +148,15 @@ export class BazaarClient implements BazaarApi {
...
+        : `/projects`;
 
-    return await this.fetchApi
-      .fetch(`${baseUrl}/projects`)
-      .then(resp => resp.json());
+    const data = await this.fetchApi.fetch(`${baseUrl}/${url}`);
```
Comment (freben) [inline plugins/bazaar/src/api.ts:158]: a hint here - the backend uses `router.use(errorHandler())` which captures backend errors in a nice, rich, detailed way. To make the client leverage this, you can add a line right here which does

```ts
if (!data.ok) {
  throw await ResponseError.fromResponse(data);
}
```

(where `ResponseError` is in the `@backstage/errors` package)

This ensures that any error details in the response body are captured and available to the caller. Especially the `<ErrorPanel>` component, which lets the user see all of those error details fully and better understanding what went wrong exactly where.

File: plugins/bazaar/src/components/BazaarOverviewCard/BazaarOverviewCard.tsx [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+      catalogEntities.value || [],
+    );
+
+    if (unlinkedCEntities) {
+      setUnlinkedCatalogEntities(unlinkedCEntities);
+    }
+  }, [bazaarProjects, catalogEntities]);
+
+  if (catalogEntities.error)
+    return <Alert severity="error">{catalogEntities.error.message}</Alert>;
```
Comment (freben) [inline plugins/bazaar/src/components/BazaarOverviewCard/BazaarOverviewCard.tsx:118]: as mentioned above, if you use `<ErrorPanel error={error} />` here (and below) instead, the user will be able to see a lot of details about why things went wrong, in a way that's consistent with the rest of Backstage

**Review by freben — COMMENTED** [review]
Thanks, nice! Just a remaining question about the API design, then we're ready!

File: plugins/bazaar-backend/src/service/router.ts [hunk 7]
```diff
@@ -92,8 +92,14 @@ export async function createRouter(
     response.json({ status: 'ok', data: data });
   });
 
-  router.get('/projects', async (_, response) => {
-    const data = await dbHandler.getProjects();
+  router.get('/projects', async (request, response) => {
+    const limit = request.query.limit?.toString();
+    const order = request.query.limit?.toString();
```
Comment (freben) [inline plugins/bazaar-backend/src/service/router.ts:97]: copy-paste error - should be "order"

**Review by freben — COMMENTED** [review]
Alright, one tiny bug left and then we're done!

**Comment by suuus** [comment]
Thank you for contributing during Hacktoberfest 🍂! You can claim the Backstage Hacktoberfest Holopin at https://bck.st/hacktoberfest-holopin 🙌🏻

**Pull request merged**